### DISCUSSION TITLE
Extendible Java-Enums in C#!

### DISCUSSION THREAD

**Discussion by najak3d** [dotnet/csharplang#5889] [Language Ideas] [answered]
Everywhere I go, I bring my tedious invention/implementation of Struct-Enums, that emulate Java-Enums, except are also Extendible (awkwardly).   It buys us the convenience/safety of a hard-coded Enum member in user code, in lieu of using "raw string key names" (which are too loose and problematic, especially for larger scale apps).

What we do is create a struct that is essentially an Enum-like smart-integer Index value, that is passed around like a flyweight.  The struct type itself holds an array of instances of this enum, referenced by Index.   In code, ToString() shows the "Name", and, like Java-Enums, you can add on as much additional data or meta-data to enum value as you desire.   In the end, most of your code simply references these values by the StructEnum Index value. 

By practice (but no mandatory) we then ALSO define all base valid members on the struct itself, so that the usage notation mimics that of a standard C# Enum.

**Example Code:**
```c#
public struct MapLayerName
{
   static private string[] s_Names = new string[MaxEntries];   // NOTE: more tables like this can be added as you desire.

    // ENUM MEMBERS/OPTIONS GO HERE - accessed like Enum values
    static public MapLayerName _NULL = new MapLayerName("_NULL");
    static public MapLayerName Roads = new MapLayerName("Roads");
    static public MapLayerName Rivers = new MapLayerName("Rivers");
    static public MapLayerName Railroads = new MapLayerName("Railroads");

   public string Name => s_Names[_index];  // MORE properties can be added, to make to each associated table above
   private int _index;
   static private int s_NextIndex;

   public MapLayerName(string name)  
   {
     _index = s_NextIndex++; 
     s_Names[_index] = name
   }
   public override string ToString() => Name;
}
```
**Usage code then looks similar to Enums:**
```c#
public void EnableLayer(MapLayerName name)  { } // enables the map layer

public void OtherMethod()
{
   EnableLayer(MapLayerName.Roads);
}
```

There is more code not currently shown here required to make this more elegant (such as overriding '==', '!=", Equals(), GetHashCode()" as well as logic to handle errors (too many entries, duplicate entries, and Dictionary to Look-Up-By-Name, and also for Serialization/Deserialization.

My proposal is that a construct similar to this be encoded into the C# language, much like Java Enums, but Extendible, to allow the addition of new Key Names run-time.   Conflicting Names either Override existing definition, or throws an exception (depends upon [Extendible] Attribute, see below).

**Suggested Attribute:**
**[Extendible(Mode = Unique/Override)]**
    - absence of Attribute means it's sealed.  It operates like a normal Java-Enum.  Only the original define Members are valid.
    - Mode = Unique -- duplicate Name entries **Throws Exception** -- not allowed
    - Mode = Override -- duplicate Name entries allowed, but OVERRIDES the existing Entry... resulting in only one entry that contains the last registered definition.

**[Key\<string>("ID")]** - Allows alternate Key Property Name/Type, to change the default of "string Name".

==
I take my StructEnum implementation with me everywhere I go, as I find it an immensely superior alternative to using raw string name keys, which is what most apps that I see, seem to be doing.  Why?  IMO, because they don't know about this alternative that I invented.   But like most things I invent, I usually find that I'm not nearly alone (others also inventing same thing elsewhere -- like the South Park Fish-D*ck-Joke).

If this feature were implemented in C#, it would probably become widely used, like Java Enums, but better (because they are extendible, and so are useful for Key-Name situations, where that list must be extendible by other assemblies/users).

==
**NOTE:**  This could also satisfy Issue #3491, as shown in the code sample below:

Code for that would look like this:
```c#
namespace Core
{
  [Extendible("Override")]
  public EnumKey ColorName : byte 
  { 
     Options = {
       _None, 
       Black = (PackedValue = #000000FF), // auto-sets "Black" as the name.
       White = (PackedValue = #FFFFFFFF)
       Red = (PackedValue = #FF0000FF)
     };
  
     public Name { get; init; }  // IMPLICIT Property -- built-in.   This is the Key value, and must be unique.
     public uint PackedValue { get; init set; } // THIS IS STORED IN A STATIC ARRAY, and is NOT an Instance field.
     public SKColor ToColor() { return new SKColor(PackedValue): }
     static public implicit SKColor(ColorName name) => name.ToColor();
  }
}

namespace OtherNameSpaceInAnotherAssembly
{
  public EnumKey ColorName: Core.ColorName 
  {    // This is ONLY allowed to Add Options -- nothing else.  This is creating the Core.ColorName instances, not a new instance type.
       AddedOptions = {
             Yellow = (PackedValue = #FFFF00FF),
             Blue = (PackedValue = #0000FFFF)
       }
  }
}

```

The Core.ColorName and MyNameSpace.ColorName would both map to "Core.ColorName" type -- literally the SAME TYPE.   The only thing EnumKey Extensions can do is AddOptions -- period.   Extending EnumKeys are only for sake of creating new members that emulate Enum values in usage.

https://github.com/dotnet/csharplang/issues/3491

**Comment by CyrusNajmabadi** [comment]
can you show how you would actually make extending work?

**Reply by najak3d** [reply]
@CyrusNajmabadi : wrote - "can you show how you would actually make extending work?"

Doesn't my "ColorName" example show that?   This shows how we'd use it.

Are you talking about how I would want to see it implemented by C# beneath the covers?

**Reply by CyrusNajmabadi** [reply]
You show the code you want to be able to write.  I'm asking how it would actually work.  What would it compile down to?  How would it, and other extensions interact with each other?

**Comment by CyrusNajmabadi** [comment]
> resulting in only one entry that contains the last registered definition.

What does 'last registered definition' mean?

**Reply by najak3d** [reply]
If a NameKey is already defined and is then registered again later with new properties, it adopts the new definition.  Of course this would have to be done before the Name space is "Closed" if it's set to be closed.

**Reply by CyrusNajmabadi** [reply]
What is this "closed" concept?  I don't see it in your OP.  THanks!

**Comment by CyrusNajmabadi** [comment, marked as answer]
Why can't you just do:

```c#
public class MapLayerName
{
    // ENUM MEMBERS/OPTIONS GO HERE - accessed like Enum values
    static public MapLayerName _NULL = new MapLayerName("_NULL");
    static public MapLayerName Roads = new MapLayerName("Roads");
    static public MapLayerName Rivers = new MapLayerName("Rivers");
    static public MapLayerName Railroads = new MapLayerName("Railroads");

   public MapLayerName(string name)  
   {
        Name = name;
   }

   public override string ToString() => Name;
}
```

Thsi seems to have exactly the same functionality as you mentioned above.    If someone wants to expose more values, they just do:

```c#
class MoreLayerNames
{
    public static readonly MapLayerName Whatever = new MapLayerName("Whatever");
}
```

You say you want the benefit of java-enums... but the point of java enums are that they are not extendible.  If you want java-enums with extendibility... you can just use a class (or a record).

**Reply by najak3d** [reply]
You suggesting use of a class, instead of struct -- turns out to be the best answer, I think.

I used a struct due to unfounded fear that too many references adds to GC load -- turns out that impact is negligible.

Therefore, I will probably be refactoring my StructEnums to ClassEnums.   I'll post back here, if I find out more, or make a move in that direction.

**Reply by najak3d** [reply]
I'd still prefer that structs allow us to derive for adding Methods and static additions, without actually polymorphing any existing methods, or changing the data/field layout.   Structs make better flyweights, IMO.... it's just a specific-typed integer, after all.. much like a standard enum.

**Reply by CyrusNajmabadi** [reply]
I feel like the proposal is a bit 'all over the place'.  It started as a way to just add values to an enum, and it has morphed to different, and very domain specific needs (for example, speculative perf constraints based on past experience).  It's really hard to get a sense of what is being asked for here, and what would actually be a satisfactory solution.

> I'd still prefer that structs allow us to derive for adding Methods and static additions

This sounds like a sub-request of the roles/shapes we're doing.  Extensions would be able to add this without the need for derivation here. :)

**Reply by najak3d** [reply]
Extensions proposal seems promising.

Yes, my proposal is "all over the place" because I'm flexible on what I get.   All I know is that C#'s limitations have seemingly unnecessarily made my task more difficult than it should be (for example, not letting me extend a struct in any fashion at all).

I'd be glad to see ANY move in a direction that makes structs more usable, or adds a new EnumKey concept that behaves in a fashion similar to what I'm doing now, but better.

I'm going to assume this proposal is dead-in-the-water, for now.   My next steps in coding may be to refactor my StructEnums into StructClasses, so that I can benefit from derivation/inheritance.

**Reply by CyrusNajmabadi** [reply]
> Yes, my proposal is "all over the place" because I'm flexible on what I get. All I know is that C#'s limitations have seemingly unnecessarily made my task more difficult

I feel liek a lot of your complaints are due to self inflicted choices.  This topic started because you felt you had to avoid an existing feature that suitably solved the problem at hand.  This has happened a few times (for example, not using features available in object-initializers, not using analyzers, etc.).  When this happens, you've asked for a new feature to address this, while not necessarily doing an effective job explaining why things as they exist are not suitable to addres the problem, but insisting that if you had this new feature you'd now be ok with that.

> All I know is that C#'s limitations have seemingly unnecessarily made my task more difficult than it should be (for example, not letting me extend a struct in any fashion at all).

This is somewhat frustrating.  The linked issues describe many reasons why struct-extension is problematic.  Painting it as "unnecessarily made my task more difficult" puts out the message from you that not doing these things was somehow a mistake, which has led only to your current problems.  As opposed to potentially something that was not done for either good reason, or substantial difficulty.

I truly ask that you either engage with people in some of our other forums, or you spend some time reading up on these issues because jumping to these immediate conclusions

**Reply by najak3d** [reply]
I think you are off base here.  The "Extensions/Roles" proposal is going to essentially permit the same thing -- adding both static and instance members to enums!...     So if Extensions/Roles can do it fine -- then why not just allow it to be done the "direct fashion" too?   The direct fashion can be compiled to produce the exact same thing -- it's not really modifying the existing enum, but rather just extending it.

So why not just allow direct Enum derivations, which are only permitted to extend, but not modify?   This would be more direct and easy to use than extensions, which are mysterious by nature (pulled in by a "using" at the top of your class file).

**Reply by najak3d** [reply]
So my statement of "unnecessary limitations on structs", I believe is accurate.  I do believe it was a mistake not to make structs "extendible" to the extent "Roles/Extensions" is already proposing to do (via the mysterious extensions mechanism).

**Reply by CyrusNajmabadi** [reply]
> So if Extensions/Roles can do it fine -- then why not just allow it to be done the "direct fashion" too?

I said why.  Because we would then have multiple ways of doing the same thing instead of one system that subsumes both.

**Reply by CyrusNajmabadi** [reply]
> it's not really modifying the existing enum, but rather just extending it.

Yes... it's *extending* it.  Hence why I think this can be subsumed by our efforts to expand on what extensions can do.

**Reply by CyrusNajmabadi** [reply]
> So my statement of "unnecessary limitations on structs", I believe is accurate. 

No, it's not accurate.  The work to make this possible is complex and requires a lot of thought into the enormous implications not just for the upcoming feature but the entire future of C#.  It's not unnecessary to be slow and methodical on that sort of thing.  It's *extremely* necessary to prevent mistakes and issues that we then have to live with for potential decades.  Not including things is important and necessary so that you don't end up with problems and complexity that you regret later.

**Reply by najak3d** [reply]
@CyrusNajmabadi  - you can already extend Classes to via Extensions - so why do we allow anyone to derive from and extend a class?  Isn't that TWO WAYS to do something?

Extensions should be the near "last resort" method for doing something, where another method doesn't make better sense.  Such as for Fluent-Pattern -- extensions make sense, because it adds methods that DO NOT MAKE SENSE for the class itself to implement.  The methods added ONLY APPLY TO COMPOSITION context, and no where else.  Therefore, Extensions make sense for that context.

However, for extending a struct in the same manner as you would extend a class - via deriving a new class -- is the most sensible direct way to do it!...      Extensions will be able to do nearly ANYTHING -- because of it's nature.  That doesn't mean you should use it for EVERYTHING!   It should be the near last resort -- as it's the least straight-forward way to do things, so should be used sparingly, and mostly only in certain contexts, with few exceptions (like Linq expressions).

**If I want to create a struct derivative to add members to it -- by far, hands down, no contest -- it makes the most sense to simply enable us to derive from the struct and do this.**   It really surprises me that this notion isn't already obvious to you (and that you instead think Extensions is an equivalently good solution).

**Reply by najak3d** [reply]
Can you tell me one-danger that you perceive for enabling structs to derive and extend other structs in the direct manner?  And why do you view it as less-dangerous to be able to do this same thing via Extensions?

It seems to me that you can't argue for Roles/Extensions to be able to "add instance/static members to structs", but then also say that this is dangerous to do it the direct/clear way of doing the exact same thing.

The added advantage of enabling direct struct derivations of this nature, is that the base struct can start using "protected" fields/members, to enable the derivation to call -- which will NOT be available for Extension Methods to call.   So Extension Methods are also inferior with regard to this capability.

**Reply by CyrusNajmabadi** [reply]
> It seems to me that you can't argue for Roles/Extensions to be able to "add instance/static members to structs", but then also say that this is dangerous to do it 

I never said it was dangerous to do. Please stop propping up strawmen.

**Reply by CyrusNajmabadi** [reply]
> you can already extend Classes to via Extensions - so why do we allow anyone to derive from and extend a class? Isn't that TWO WAYS to do something?

First: Extending a class supports adding state. Something your struct-insurance proposal does not.  So they're not comparable.  If you were to allow adding state, I would say using inheritance would be acceptable. 

Second: yes, when it come to classes, if you are adding functionality in the same manner that is discussed here I would absolutely be using extensions and not inheritance.

**Reply by CyrusNajmabadi** [reply]
> Extensions should be the near "last resort" method for doing something, 

This is your own pov, but it is not a trueism. We use extensions commonly, not as a last resort, but because we view them as the right abstraction for the problem. For example, we commonly used them to augment a type *we own* with functionality because we think an instance member is actually undesirable. This allows us to keep the actual surface api if the type itself minimal, while still making it pleasant to work with instances.  It helps us design and make clear what the smallest set of functionality is truly necessary that needs access to state (or privates). 

Extensions are in no way last resort. They are an abstraction tool for separating out responsibilities and code locations.

**Reply by CyrusNajmabadi** [reply]
>  It really surprises me that this notion isn't already obvious to you

I've given you my reasoning. I'm not sure where your surprise stems from. 

> it makes the most sense to simply enable us to derive from the struct and do this. 

It is not simple for one. To start with, the runtime literally does not support this.  For another, as mentioned a few times, adding a feature that seems like it should be able to do something (like add state), but which doesn't makes ideas like that less palatable. And, finally, as mentioned, we are already working on an approach that subsumes the set of functionality we are adding here. So it would be time and effort taken away to add support for a more limited set of cases that then is subsumed by the broad set of cases coming. That is needless complexity at the end.

**Reply by CyrusNajmabadi** [reply]
> the base struct can start using "protected" fields/members, to enable the derivation to call 
> ...
> Can you tell me one-danger that you perceive for enabling structs to derive and extend other structs 

Again, i didn't say anything about danger.  However, because you brought it up, i'm happy to go into that topic.

Before i do though, i want to point out that i find this style of conversation pretty offputting.  Time and time again I've seen you jump your interest from one idea to something else entirely with the view that now *this* proposal will be the solution to the things that ail you.  You push for something intensely, until you discover that what you were asking for doesn't actually solve your issue, or that your set of criteria you care about isn't necessarily even relevant or accurate anymore. At that point you find something else and it gets all your attention and insistence.  WRT to the discussion goign on right now, I'm not even sure why you even want a struct once you learned that you could solve your problem with classes.  Previously you mentioned you needed it because of perf reasons.  However, the description of your use case was one that would be fine for performance without ever needing a struct to solve the problem.  As others have mentioned, lang design will absolutely consider what options you already have.  And if you have perfectly reasonable options, and you cannot elucidate why those options are not viable and only this path addresses things, then we're highly unlikely to take this path.

Ok, back onto 'protected' and 'danger'.  First off, this is the first you mentioned about 'protected' being available in a derived struct.  Previously you only mentioned adding members.  But ok, let's look at this new requirement you're talking about.  First, remember that 'protected' members (and virtual for that matter) are not legal in structs.  But that's ok, let's assume we can relax that and allow a 'protected' field in a struct.  What would be the consequences of that?  Well, right off that bat it opens up an extremely dangerous ability for someone to access your protected members simply by deriving from your struct and then casting an instance of your struct to the sub-struct.  e.g.

```c#
struct Base
{
     protected int X;
}

public Base MyAPI() { return new Base(...); }

// then in different assembly.

struct Other : Base
{
    public int GetX() => this.X;
}

var other = (Other)MyAPI();
```

This is not a problem with classes as classes actually carry around a pointer to their type information, so the runtime can actually check if that cast is legal and prevent it from happening.  Structs have no way to do that (that's literally the idea of a struct, it's just the data) so you'd be sunk here.  The only way to prevent this would be disallow downcasting with structs.  But then that would break one of the core reasons to use derivation in the first place.  What good is allowing derivation if you can't take a base struct and then cast it to a derived struct to get at those members you care about?   So, you either restrict downcasting, or you disallow protected.

This is an immediate and severe (and dangerous) issue with this concept right off the bat.  And, as per above,tThe point of telling you this is not to say that this idea does not have merit.  Rather, the point is to demonstrate to you that language design cannot be *haphazard*.

A common pattern i've seen in your posts is you grabbing an idea with a ton of enthusiasm, claiming it is awesome, outright saying that it both needs to be done asap, and it's entirely unnecessary that we haven't done it already and that us not doing it is hurting you.  But these types of problems demonstrate deeply why it's not only not trivial to provide such features, but that:

1. if we are not careful we could end up with something actively bad and dangerous.
2. even if we are careful, we'll end up with something non-sensical.  e.g. derived types that do not work at all like how derived types work elsewhere.

I think it's great you have this interest and that you are willing to fight for these ideas.  What i strongly ask though is that you curb your insistence that a particular idea is the 'obviously' right way to go, and that it would be trivial to just add these features without any problems whatsoever.  There are deep and challenging problems in this space, and we are focusing on areas that we believe maximize the value while avoiding those very issues.  

Instead of browbeating with posts like: "It really surprises me that this notion isn't already obvious to you" and "by far, hands down, no contest -- it makes the most sense" to do your ideas.  It would be great if you approached these ideas demonstrating a willingness to converse and learn more about the topic (including the pros/cons) with the team and community to help find a potential idea that really is great, instead of just assuming the one you just stumbled upon is already there without any effort put into validating that.  Thanks!

**Reply by najak3d** [reply]
Here's the rebuttal of why those "dangers" seem nonsense to me:

1. Existing structs all have 'private' members only -- so there is zero risk of someone abusing an existing struct in this fashion, of accessing protected members.   The only protected members that will exist will be future cases, using latest C# feature, to purposefully allow derivatives to have access to these fields/members.   There is no danger here of someone doing something bad.

2. The actual "type" of the derived struct is the SAME as the base struct.  Just view it as a "struct with Extension methods enabled".  This would work the exact same as the new "Roles/Extensions" proposal, but with two distinct advantages:
a. Struct overrides can take advantage of 'protected' access, while the new Extensions proposal will be limited to public access only.
b. Struct override obtain these Extensions directly/simply - by casting the struct to the new struct type -- this simply enables the extensions of the new struct type.  Period.  It would be like saying "for this struct instance, enable the Extension methods".  

So there is zero added risk here, compared to the new "Roles/Extensions" proposal.   Yet it provides at least two very distinct added benefits.

==
You seem to still think that "I am bull-headed and don't listen" or "not interested in discussion" - but the evidence you just laid out of "me quickly changing direction" is evidence of the opposite.   I hear arguments, and then modify my proposal to be something that appears to have MORE MERIT -- so I change my proposal.  The Proposals are just maturing -- based on feedback and additional thought. 

For example, I've backed off on the importance of making "StructEnums" due to your feedback, and some testing I just did, to see if "ClassEnums" were acceptable -- and I concluded that they are.    Which means there is less of a need for this current proposal of "Exendible Java Enums" because half-of-the-pains I'm experiencing with StructEnums has to do with the unnecessary restrictions placed upon Structs (i.e. I'm not trying to change it's type, but only trying to Extend with new members).

**Reply by CyrusNajmabadi** [reply]
> The only protected members that will exist will be future cases, using latest C# feature, to purposefully allow derivatives to have access to these fields/members. There is no danger here of someone doing something bad.

There is.  A user would very normally intuit that a protected member in a struct woudl have hte same protections as in a class.  If htis violates that expectation that's a problem that will lead to confusion and potential serious issues.  We cannot handwave that away by saying "it's new".

**Reply by CyrusNajmabadi** [reply]
> So there is zero added risk here

I literally outlined the risk.  Someone makes a member protected, expecting the same protection you get with classes, and discovers that that is not the case.

**Reply by CyrusNajmabadi** [reply]
> I'm experiencing with StructEnums has to do with the unnecessary restrictions placed upon Structs 

Again, they're not unnecessary.  Taht was exactly the thing i pointed out you keep saying, but these examples show substantive reasons and concerns about.  Please stop being so dismissive and unwilling to accept core concerns we have about the language and concepts like intuition and around corresponding analogous features.

> but only trying to Extend with new members).

This is not true.  YOu're not just trying to extend with new members.  You're also trying to get access to protected members.  

> I'm not trying to change it's type, 

Then that's exactly the domain of extension members.  They do not change the type, but they add supplementary members that people can use.

**Reply by CyrusNajmabadi** [reply]
>  I hear arguments, and then modify my proposal 

I haven't seen modifications, i've just seen jumping to entirely otehr proposals.  And this comes after days of you steadfasting insisting that the proposal you are currently on is both obviously the thing that should be done.  This keeps happenign until we pry things apart to discover that isn't the case.  THen you jump to something else and the cycle continues.  

As i've mentioned, it would help a lot if you could jump to a different feature but *not* just insist it's absolutely teh best and the right hting forward, but instead work to understand the concerns and issues involved in teh space with a headspace to find a good solution that meets the needs and requirements of the entire ecosystem, not just yourself.

**Reply by CyrusNajmabadi** [reply]
> For example, I've backed off on the importance of making "StructEnums" due to your feedback, and some testing I just did, to see if "ClassEnums" were acceptable -- and I concluded that they are.

I'm really hoping you read back on your posts and look at them again with the lens of where you are now.  Recall that you were entirely insistent that what you were saying was super important and it was unfathomable to you that we both did not agree with you, and that we weren't jumping to make this happen.  And yet, a couple days later, you've entirely dropped that line, but are still makign the same sorts of arguments with the new interesting topic du-jour.  If you keep making these same sorts of arguments (again, that something is super critical, and obvious, and none of the concerns are relevant or acceptable to you), then you're going to be less and less effective because those arguments are not turnign out to be true, and it's sending the message that you're not willing to accept alternative points of view.

**Reply by BrainSlugs83** [reply]
> You suggesting use of a class, instead of struct -- turns out to be the best answer, I think.
> 
> I used a struct due to unfounded fear that too many references adds to GC load -- turns out that impact is negligible.
> 
> Therefore, I will probably be refactoring my StructEnums to ClassEnums. I'll post back here, if I find out more, or make a move in that direction.

If you look at some of the existing patterns, you might benefit from this:
https://github.com/Azure/azure-libraries-for-net/blob/master/src/ResourceManagement/ResourceManager/Region.cs

It has some useful bits like making the constructor private (maybe protected is better for your case), overriding GetHashCode, adding other equality checking operators, making the members `readonly` (because otherwise someone can reassign new values to your "enums"!) -- and using a concurrent dictionary, in case the same value gets created twice.

**Comment by najak3d** [comment]
Our intent is for these to be lightweight structs, not class references.   Classes work better because you can inherit, and so it's easier to add more values to your namespace.   With structs, it's awkward.    I'd like EnumKeys to be struct-based flyweights, where their index value is used to look up the full definition (even it's just a name).

If we could make Structs derive from other structs ONLY for adding static members -- That would be very helpful to me.  But I can't and so it's awkward.

The other awkwardness of using Class/Structs to emulate the enum, is that these classes have other Properties that show up statically, such as "ReferenceEquals" and our methods used for Registration -- so these pollute the Intellisense list when you are trying to select a member name.

At least with the base class (vs struct) approach, we can also have it know how to serialize itself to JSON, etc, and then you could just derive your other ClassEnums from this base class, and gain some benefit like that.   BUT, it was our intent to make this a struct, not a class.   Thinking about it more now -- maybe making it a class would alleviate of our headaches, all for the cost of passing around references, vs. this simple 1-integer struct.

If we could classify those ClassEnum instances in a special way that tells GC to ignore it -- never to dispose them... then we wouldn't have to be concerned for adding GC load, nor increasing risk of a GC0 hiccup.   For these reasons, I steered clear of Classes, and made it work with pure enums - because they incur almost no cost to the GC.    The GC has to do work each time you pass a reference, because it has to track whether or not some other party held onto the reference.   Object counts is not the real enemy in GC, but rather the total "reference count".     1000 objects with 1 million references to those objects is a lot heavier weight than 1000 objects with only 1000 references to those object.   My StructEnums achieves the latter efficiency --  1 object for each Enum def - - but ONLY ONE STATIC REFERENCE to each.

My experience with GC and hiccups is over a decade old -- so maybe it's less of a concern now.  Our apps run on mobile devices, where processing power is substantially less than a typical PC, and so I remain vigilant about conserving on work load that we give to the GC.

I could change this proposal to "Allow Structs to be Overridden to add Static members", so that we could extend Enumerations in the fashion you suggested here -- then we've got the best of all worlds.

If we could derive from a struct for static members only -- holy moly -- that would be very nice.  Maybe I should rewrite this whole proposal to be that instead.  It would make what we do now, a lot less awkward, without turning our solution into a something that involves passing around more reference types.

**Reply by CyrusNajmabadi** [reply]
> Our intent is for these to be lightweight structs, not class references.

Why is this your intent?  

>  I'd like EnumKeys to be struct-based flyweights, where their index value is used to look up the full definition (even it's just a name).

This seems to be working backwards from a solution to particular requirements.  What's the reason to encode things in that way when it's trivial to store data in a class instead?

> The other awkwardness of using Class/Structs to emulate the enum, is that these classes have other Properties that show up statically, such as "ReferenceEquals" and our methods used for Registration -- so these pollute the Intellisense list when you are trying to select a member name.

This has nothing to do with this proposal.  Intellisense is not hte language, nor is the language intellisense.  Nothing about your enum proposal, for example, says taht these do not derive from object.  As such, they would have all these symbols inherited from object.

**Reply by CyrusNajmabadi** [reply]
> all for the cost of passing around references, vs. this simple 1-integer struct.

References are as cheap to pass as a 1-integer struct.  So i'm not sure what yuo mean here.

**Reply by CyrusNajmabadi** [reply]
> If we could classify those ClassEnum instances in a special way that tells GC to ignore it -- never to dispose them... 

I don't understand this statement.  These are statics.  They would never be disposed.  You don't need to mark anything.

> then we wouldn't have to be concerned for adding GC load, nor increasing risk of a GC0 hiccup. 

There are no continual allocations here.  You are simply defining N specific instance values.  There is no GC load or GC hiccough concern here.

> My experience with GC and hiccups is over a decade old -- so maybe it's less of a concern now. Our apps run on mobile devices, where processing power is substantially less than a typical PC, and so I remain vigilant about conserving on work load that we give to the GC.

We're not going to add such a feature based on speculative perf concerns.  :)

**Reply by najak3d** [reply]
It's not speculative.  When we wrote the "Visual3D.NET Game Engine" in 2009, we tested the GC thoroughly, because it's vital to avoid hiccups.  And what impacted GC hiccups the most was "# References" more than "# Objects".   When we gave it more References to track, that became the worst part of the workload.   Maybe it's better now; I haven't retested this in over a decade.

These enum types should be structs.  And there isn't really a good argument for not allowing Struct derivations to "add methods/properties" so long as the derivation does not modify the field-layout/composition.    Right?   So I'm now asking for that instead.

**Reply by HaloFour** [reply]
Structs would certainly perform worse here.  You are introducing a level of indirection and incurring the cost of copying where instead you would have a very small number of allocations that are made once up front and very cheap copy costs.

**Reply by CyrusNajmabadi** [reply]
>  I haven't retested this in over a decade.

This is speculative then.  We'd need modern data on things like .net core showing that this is an issue.  If perf is a motivator here, then we need to establish that there's actually a problem.  Otherwise, we can't actually measure and know if we've delivered a solution that solves that :)

> These enum types should be structs. 

I don't see a reason why, not if there are suitable lang solutions today that already work.  

> Right? So I'm now asking for that instead.

Can you update your proposal then?

**Reply by BrainSlugs83** [reply]
> > all for the cost of passing around references, vs. this simple 1-integer struct.
> 
> References are as cheap to pass as a 1-integer struct. So i'm not sure what yuo mean here.

References are generally cheaper than structs for passing -- that's the point of references. -- Also won't hit the GC because you only create the instances once at start-up and never lose scope of them since they should be `public static readonly` values.

**Reply by CyrusNajmabadi** [reply]
They still hit the gc as they must always be examined when the gc runs.

**Comment by najak3d** [comment]
I just opened a new Idea/Discussion, which attacks this from a different vantage point, of C# enabling us to create these Java-Like-Smart-Enums that are also extendible -- and also remain as Structs.

https://github.com/dotnet/csharplang/discussions/5890

**Comment by najak3d** [comment]
If there were a way to classify some objects as "non-disposable" such that the GC simply would ignore these references as they are passed around, and never ever try to see if they can be disposed -- then class-based-enums would be more attractive/efficient.

So if this were possible, then I could turn that feature on for Class-Enums, and pass around these references without incurring any load to the GC, and adding no hiccup risks.

**Reply by HaloFour** [reply]
GC is only involved if you are constantly making allocations.  Having a set of instances does not result in constant allocations and puts zero additional pressure on the GC.  Java enums are implemented as reference types because passing reference types around is cheap and each value is only ever allocated once.

**Reply by najak3d** [reply]
@HaloFour  - that's how we *used* to think of it, until we tested it, back in 2009.  We thought GC hiccups were triggered by "too many objects", and didn't realize that a large part of the equation was "references to those objects"...  The GC has to do detective work to figure out if an object on the heap has no more references to it...   and the number of references plays a big role in that workload.    1,000 objects with 1000 references is a lot cheaper than 1,000 objects with 1 thousand references each.  Maybe that was an old issue that has been resolved.   We tested it and came to that conclusion from test results.  But this was back in 2009.   Maybe I should do it again now, and see what comes of it.

**Comment by najak3d** [comment]
Test results are in.   And my concern about "#References" adding to GC load is no longer a thing (maybe never was, can't tell for sure since our prior conclusions/tests are from 13 yrs ago).

I just ran an object creation/disposal stress test at App startup, in two different modes.

Case #1:
1. Create 10,000 raw objects in a list.
2. Create 10,000,000 other objects that each reference an instance of the 10,000 raw objects in 1.   (each object in 1 has 1,000 references each).

Case #2:
Same as 1, but those 10,000,000 other objects reference null instead.

Then I measured for GC hiccups (assuming that any hiccup that occurred was from GC - which it typically is in a simple scenario like this).

Ran both for about 1 minute, to gather metrics.

End Result -- Case \# 1 only took 5.5% longer than \# 2.   So despite adding 10 million references alongside these 10 million objects... it only added 5.5% of work load to the GC.   That is negligible.

And so the impact of Object creation, destruction, and compressing the heap -  has about 20x the impact on the GC as does the same amount of References. 

And so I'm going to stop worrying about Reference Counts entirely.   It's negligible in comparison to object creation/destruction/heap-compressing that is already being done by the GC in these cases.

And so ... maybe I'll reconsider using a Class instead of a Struct for my StructEnums solution -- which then will enable it to be a more elegant solution.   (Derived ClassEnums can derive from the base class, which provides the infrastructure for managing the Names Lists, and memberships, error handling, and such). 

Here are my test results -- Left Side is Case #2,   and Right Side is Case #1 -- took 5.5% more in hiccups.
(I could make it hiccup about once per 3 seconds.)

![image](https://user-images.githubusercontent.com/10914515/156974282-7c2441e3-d575-421f-a0e8-4487e192a22e.png)

Here is the code:
```c#
static public List<object> s_Objects = new List<object>();
static public List<ObjRef> s_Refs = new List<ObjRef>();
static public List<string> s_Hiccups = new List<string>();
static private int NumObj = 10000;

public class ObjRef
{
	public object Obj;
}

static public void Test()
{
	int longestHiccup = 0;
	int beginTicks = System.Environment.TickCount;
	int iterations = 0;
	int HiccupsTotal = 0;

	while (true)
	{
		iterations++;

		int tickStart = System.Environment.TickCount;
		int hiccup = Allocate();
		s_Refs.Clear();
		int elapsed = System.Environment.TickCount - tickStart;

		if (hiccup > longestHiccup)
		{
			longestHiccup = hiccup;
		}
		if (hiccup > 50)
		{   // Hiccup.
			int ticks = System.Environment.TickCount - beginTicks;
			s_Hiccups.Add(ticks + ":" + iterations + " - Hiccup: " + hiccup);

			HiccupsTotal += hiccup;
			if (s_Hiccups.Count >= 40)
			{ }
		}
		else { }
	}
}
static private int Allocate()
{
	int hiccup = 0;
	int tickStart = System.Environment.TickCount;
	s_Objects.Add(null); // NOTE:  s_Objects[0] == null!! (for TEST CASE #2 below)

	for (int i = 0; i < NumObj; i++)
	{
		s_Objects.Add(new object());
	}

	int ticksNow = System.Environment.TickCount;
	int elapsed = ticksNow - tickStart;
	if (elapsed > hiccup) hiccup = elapsed;

	for (int j = 0; j < 1000; j++)
	{
		tickStart = ticksNow;

		for (int i = 0; i < NumObj; i++)
		{
			//s_Refs.Add(new ObjRef() { Obj = s_Objects[0] }); // TEST CASE #2 - no references
			s_Refs.Add(new ObjRef() { Obj = s_Objects[i] }); // TEST CASE #1 - 10 million references
		}

		ticksNow = System.Environment.TickCount;
		elapsed = ticksNow - tickStart;
		if (elapsed > hiccup) hiccup = elapsed;
	}

	return hiccup;
}
```
